subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PrecompiledGroovyPluginsPlugin.java

```diff
@@ -0,0 +1,86 @@
+import org.gradle.plugin.devel.GradlePluginDevelopmentExtension;
+import org.gradle.plugin.devel.plugins.JavaGradlePluginPlugin;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+class PrecompiledGroovyPluginsPlugin implements Plugin<Project> {
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().apply(GroovyBasePlugin.class);
+        project.getPluginManager().apply(JavaGradlePluginPlugin.class);
+
+        project.afterEvaluate(this::exposeScriptsAsPlugins);
+    }
+
+    private void declarePluginMetadata(GradlePluginDevelopmentExtension pluginExtension, List<PrecompiledGroovyScript> scriptPlugins) {
+        pluginExtension.plugins(pluginDeclarations ->
+            scriptPlugins.forEach(scriptPlugin ->
+                pluginDeclarations.create(scriptPlugin.getId(), scriptPlugin::declarePlugin)));
+    }
+
+    private void exposeScriptsAsPlugins(Project project) {
+        GradlePluginDevelopmentExtension pluginExtension = project.getExtensions().getByType(GradlePluginDevelopmentExtension.class);
+
+        SourceSet pluginSourceSet = pluginExtension.getPluginSourceSet();
+        FileTree scriptPluginFiles = pluginSourceSet.getAllSource().matching(PrecompiledGroovyScript::filterPluginFiles);
+        if (scriptPluginFiles.isEmpty()) {
+            return;
+        }
+
+        List<PrecompiledGroovyScript> scriptPlugins = scriptPluginFiles.getFiles().stream()
+            .map(PrecompiledGroovyScript::new)
+            .collect(Collectors.toList());
+
+        declarePluginMetadata(pluginExtension, scriptPlugins);
+
+        TaskContainer tasks = project.getTasks();
+
+        Provider<Directory> javaClasses = tasks.named("compileJava", AbstractCompile.class).flatMap(AbstractCompile::getDestinationDirectory);
+        Provider<Directory> groovyClasses = tasks.named("compileGroovy", AbstractCompile.class).flatMap(AbstractCompile::getDestinationDirectory);
+
+        TaskProvider<PrecompileGroovyScriptsTask> precompilePlugins = tasks.register(
+            "compileGroovyPlugins", PrecompileGroovyScriptsTask.class, scriptPlugins,
+            pluginSourceSet.getCompileClasspath(), javaClasses, groovyClasses
+        );
+
+        TaskProvider<JavaCompile> compilePluginAdapters = tasks.register("compilePluginAdapters", JavaCompile.class, t -> {
+            t.dependsOn(precompilePlugins);
+            t.setSource(precompilePlugins.flatMap(PrecompileGroovyScriptsTask::getPluginAdapterSourcesOutputDir));
+            t.setDestinationDir(precompilePlugins.flatMap(PrecompileGroovyScriptsTask::getAdapterClassesOutputDir));
+            t.setClasspath(pluginSourceSet.getCompileClasspath());
+        });
+
+        pluginSourceSet.getOutput().dir(precompilePlugins.flatMap(PrecompileGroovyScriptsTask::getPrecompiledGroovyScriptsOutputDir));
+        pluginSourceSet.getOutput().dir(compilePluginAdapters.flatMap(JavaCompile::getDestinationDirectory));
```

Comment:
Here's the feature I was thinking of: https://docs.gradle.org/6.1/release-notes.html#compilation-order

I think what we want to do here is define a new `SourceDirectorySet` as part of the "pluginSourceSet" (`SourceSet`) that overlaps with `src/main/groovy` with a filter on `*.gradle` files.  The new `SourceDirectorySet` will be compiled by the precompiled plugin compile task (compileGroovyPlugins).  The dependency between the precompiled plugin compile task and compileJava/compileGroovy need to be done through the classesDirectory of the other SourceDirectorySets.

I think we might be able to make the plugin adapters compile as part of the regular JavaCompile task if we generate them as a separate task instead of as part of compileGroovyPlugins.  WDYT?

Reply:
I was able to model the classpath dependencies using source set outputs finally.
It was not trivial because Java source set is retrieved like this:
`SourceDirectorySet javaSource = pluginSourceSet.getJava();` while for Groovy one, I had to do this: `SourceDirectorySet groovySource = new DslObject(pluginSourceSet).getConvention().getPlugin(GroovySourceSet.class).getGroovy();`